### DISCUSSION TITLE
How to use table inside table?

### DISCUSSION THREAD

**Discussion by jlei523** [TanStack/table#2294] [General]
I'm using the latest react-table version.

I have a table with rows that expand into more rows:
![image](https://user-images.githubusercontent.com/3815607/81506813-3b860a00-9323-11ea-87ce-05521b36be9e.png)

Expanded state:
![image](https://user-images.githubusercontent.com/3815607/81506815-43de4500-9323-11ea-8150-f12ad6da2657.png)

How can I recreate this?

Should I put another instance of react-table inside each row that expands? How can the sub-table instance share the same columns as the original table?

Thank you for your work Tanner!

**Comment by BrianMitchL** [comment]
If you add data to the `subRow` property on your data object, an expanded row using the `useExpanded` hook should display that data.

I have a use-case that I'm trying to work through right now where I'd like to fetch data for an expanded row when expanding that row, but without needing to re-supply and build all of the markup. Pre-fetching all of the subRow data for a fully controlled/manual table is too expensive from a backend perspective. Haven't figured out a good solution yet...

**Reply by raikusy** [reply]
Hey, I'm looking for a similar solution. Did you come up with anything?

**Comment by dbertella** [comment]
The above is possible to obtain with a single table but it will be very painful to maintain, basically I'm doing the same and I'm just using subRow and useExpanded as mentioned before, the problem are footers. So I precompute all the totals at the beginning and all the footers are just normal rows and I explicitly mapping them as footer (giving them a different style) code it's quite ugly but if you play with it you can make it. 

I was looking into the same solution to have multiple tables but then they will go out of sync very easily and since you are using this prepareRow function that mutates the rows it's very hard I believe to extrapolate pieces and work with it. 

If you find any solution please share it! I'll be very interested in seeing something like the above working in a nice way

**Reply by bytesagar** [reply]
@dbertella  Do   you have any references?

**Comment by jlei523** [comment]
Maybe @tannerlinsley has a good way to do this?

**Comment by vasanthi-d** [comment]
@jlei523 - were you able to find a solution to this? I have a similar use case to build.

**Reply by jlei523** [reply]
I ended up not using React-table for this. Just divs.

**Reply by vasanthi-d** [reply]
Thank you @jlei523 for the response.

**Comment by bipin-n43** [comment]
I have a similar use case , can you please suggest me the best approach to do it as im using react-table in all other pages of project so would be better if its possible with react table, i managed to render table inside a table but i couldnt add a logic that displays choosen subtables , apparently i used state to figure out which row is clicked but it displays just one and hides all others, please suggest me an idea to hide show multiple subtables

**Comment by yerokhin-art** [comment]
Seems, [material-react-table](https://www.material-react-table.com/) which is based on this library has such feature "[Detail Panel](https://www.mantine-react-table.com/docs/guides/detail-panel)" to render different stuff, when row is expanded. I tried to research source code of that library, and I couldn't understand how does it bypass "subRows" property, because @tanstack/table, as I understand, depends on "getSubRows" property, when you implement row expanding. One more thing is that the developer of that lib is one of the devs of @tanstack/table. Maybe he could show some example how to do this?

**Reply by chie7tain** [reply]
Were you able to figure out how to add another table inside the expansion with diff headers ? Because I am trying to do the same now even the Material-react-table lib ? It would really help me out

**Reply by hanif-ali** [reply]
@chie7tain were to able to figure this out?

**Reply by raychengy** [reply]
I was able to figure out a way to display a nested table that has different headers/columns.

One method requires you to use "row.[toggleExpanded](https://tanstack.com/table/v8/docs/api/features/expanding#toggleexpanded)" as your on-click handler instead of `getToggleExpandedHandler` since that method checks for "row.[getCanExpand](https://github.com/TanStack/table/blob/a1e9732e6fc3446a2ae80db72a1f2b46a5c11e46/packages/table-core/src/features/Expanding.ts#L341C7-L341C43)". And on your UI-table implementation, you'll need to add your own nested table depending on `row.isExpanded`. Here's an example of something similar to above [codesandbox](https://codesandbox.io/p/sandbox/new-glade-frtjg9?embed=1&file=%2Fsrc%2Fmain.tsx%3A1%2C1).

Or you can overwrite the table option "[getRowCanExpand](https://tanstack.com/table/v8/docs/api/features/expanding#getrowcanexpand)" to bypass the default check of "row.subRows". This will allow you to use `getToggleExpandedHandler` for your on-click handler. You'll still need to implement on the UI-table the nested table within your row.

**Reply by adamspotlite** [reply]
@raychengy looks like this `codesandbox` link is busted - do you have another / put in a code example?

**Reply by javed24** [reply]
@raychengy same query as @adamspotlite from above. Could use a working link for the example you mentioned above.

**Comment by jhigger** [comment]
I had the same issue and I just wanna share the hack that i did to solve this problem in a table for mobile screen size in a project I'm doing.
I added an accordion and it puts a blank row at the bottom when you click it. The sub table is also shown on top of the blank row as an absolute ```<td></td>``` tag relative to the row of the main table. This new sub table can also be scrolled horizontally. I only displayed a few column headers on the row of the main table and so I excluded those column headers in the sub table using filter.

I am new to this so please excuse my code thank you.

```
{/* mobile table */}
<table className="relative w-max min-w-full md:hidden">
  <thead className="title-sm sticky top-0 z-20 bg-white font-medium text-neutral-900">
    {table.getHeaderGroups().map((headerGroup) => (
      <tr key={headerGroup.id}>
        {headerGroup.headers
          .filter((header) => filters.includes(header.column.id))
          .map((header) => (
            <th key={header.id} className="border-0 p-[8px]">
              {header.isPlaceholder ? null : (
                <div
                  {...{
                    className: header.column.getCanSort()
                      ? "cursor-pointer select-none flex gap-[10px]"
                      : "flex items-center",
                    onClick:
                      header.column.getToggleSortingHandler(),
                  }}
                >
                  {flexRender(
                    header.column.columnDef.header,
                    header.getContext(),
                  )}
                  {header.column.getCanSort() && (
                    <ArrowSwapVertical
                      className="self-end text-neutral-500"
                      size="20"
                    />
                  )}
                </div>
              )}
            </th>
          ))}
        <th className="border-0 p-[8px]">
          {/* placeholder for accordion */}
        </th>
      </tr>
    ))}
  </thead>
  <tbody className="border-y-[1px] border-gray-500 [&>tr]:border-b-[1px] [&>tr]:border-gray-500 last:[&>tr]:border-b-0">
    {table
      .getRowModel()
      .rows.slice(startRowIndex, endRowIndex)
      .map((row) => {
        return (
          <TableRow
            key={row.id}
            row={row}
            filters={filters}
            columns={columns}
          />
        );
      })}
  </tbody>
</table>
```

```
type TableRowProps<T> = {
  row: Row<T>;
  filters: string[];
  columns: ColumnDef<T>[];
};

const TableRow = <T extends object>({
  row,
  filters,
  columns,
}: TableRowProps<T>) => {
  const { open, toggle } = useAccordion(false);

  return (
    <Fragment>
      <tr className="relative bg-gray-10">
        {row
          .getVisibleCells()
          .filter((cell) => filters.includes(cell.column.id))
          .map((cell) => (
            <td key={cell.id} className="border-0 p-[8px]">
              {flexRender(cell.column.columnDef.cell, cell.getContext())}
            </td>
          ))}

        <td className="border-0 p-[8px]">
          <AccordionButton open={open} toggle={toggle} />
        </td>

        {open && (
          <td className="absolute left-0 top-[76px] z-10 w-full overflow-auto border-0 border-b border-gray-500 bg-gray-10 p-0">
            <SubTable data={[row.original]} columns={columns} />
          </td>
        )}
      </tr>
      {open && (
        <tr>
          <td className="border-0 p-[8px]">
            <div className="mb-[96px]"></div>
          </td>
        </tr>
      )}
    </Fragment>
  );
};
```

```
interface SubTableProps<T extends object> {
  data: T[];
  columns: ColumnDef<T>[];
}

const SubTable = <T extends object>({ data, columns }: SubTableProps<T>) => {
  const table = useReactTable({
    data,
    columns,
    getCoreRowModel: getCoreRowModel(),
  });

  const filters = useMemo(
    () => ["checkbox", "images", "orderId", "actions"],
    [],
  );

  return (
    <table className="relative w-max min-w-full bg-[#EBE0E859]">
      <thead className="title-sm sticky top-0 z-10 font-medium text-neutral-900">
        {table.getHeaderGroups().map((headerGroup) => (
          <tr key={headerGroup.id}>
            {headerGroup.headers
              .filter((header) => !filters.includes(header.column.id))
              .map((header) => (
                <th key={header.id} className="border-0">
                  {header.isPlaceholder ? null : (
                    <>
                      {flexRender(
                        header.column.columnDef.header,
                        header.getContext(),
                      )}
                    </>
                  )}
                </th>
              ))}
          </tr>
        ))}
      </thead>
      <tbody>
        {table.getRowModel().rows.map((row) => (
          <tr key={row.id}>
            {row
              .getVisibleCells()
              .filter((cell) => !filters.includes(cell.column.id))
              .map((cell) => (
                <td key={cell.id} className="border-0">
                  {flexRender(cell.column.columnDef.cell, cell.getContext())}
                </td>
              ))}
          </tr>
        ))}
      </tbody>
    </table>
  );
};
```

**Comment by adamspotlite** [comment]
What is the suggested way to do a `non-subRow` style render? For instance, I would love to render an arbitrary component _under_ a row, e.g. a `table inside a table`, but every attempt I've tried ends up with the second component _only with the width_ of the first cell of the parent table.

I'm currently rendering like this:

```
          
            {table.getRowModel().rows.map((row) =>
              row.getIsExpanded() ? (
                <>
                  <UITableRow<T> key={row.id} row={row} />
                  {ExpandedRowComponent && (
                    <div className="flex w-full">
                      <ExpandedRowComponent row={row.original} />
                    </div>
                  )}
                </>
              ) : (
                <UITableRow<T> key={row.id} row={row} />
              ),
            )
          )}
          ```

Where does the onus of that render belong? Inside a `UITableRow` element? Or is it better to render _two_ `TableRow` for an expanded row - one for the principal and one for the expanded component?

Appreciate the help!

**Reply by aaronleopold** [reply]
> every attempt I've tried ends up with the second component only with the width of the first cell of the parent table

I am using a nested table in one of my projects, it isn't perfect (and I haven't figured out coordinating virtualization between the nested tables yet) but hopefully this helps: https://github.com/stumpapp/stump/blob/develop/packages/browser/src/scenes/smartList/items/table/GroupedSmartListItemTable.tsx#L153-L159

The TLDR; for getting the sizing right is that I wrap my nested table in a row with one cell that has a `colSpan` of the number of columns in the parent table.

**Reply by adamspotlite** [reply]
@aaronleopold you know when you come across someone random on the internet that just has _exactly_ what you're looking for? you were that for me today! Thanks so much.

**Reply by dlibis** [reply]
@aaronleopold did you delete the conpnent? Getting 404 on dev branch

**Reply by aaronleopold** [reply]
I've updated the link, the component was relocated
